### PR TITLE
WIP Multi-server camus deploy

### PR DESCRIPTION
**Pull request by jsuresh**

### CODE REVIEW

**Comment by jsuresh** [comment]
Tests to come, thought I'd post a WIP PR that has multi-server camus implemented.

File: application.go [hunk 1]
```diff
@@ -50,6 +53,10 @@ type ApplicationDef struct {
 
 	// e.g. user@host  (no path)
 	Targets map[string]*Target
+
+	// A map of names to a list of targets. Useful for multi-server deploys. It's an
+	// error for the same key to exist in both TargetGroups and Targets
```
Comment (soulplant) [inline application.go]: Mention that the returned list are keys into the Targets map.

File: application.go [hunk 2]
```diff
@@ -96,6 +103,12 @@ func ApplicationFromConfig(isClient bool, file string) (Application, error) {
 		if !foundTarget {
 			return errMsg("No targets specified")
 		}
+
+		for name, _ := range def.TargetGroups {
```
Comment (soulplant) [inline application.go]: We should also check that all targets mentioned in TargetGroups are defined in Targets.

File: application.go [hunk 3]
```diff
@@ -116,6 +129,23 @@ func ApplicationFromConfig(isClient bool, file string) (Application, error) {
 func (a *AppImpl) RunCmd(port int) string {
 	return strings.Replace(a.def.RunCmd, "%PORT%", fmt.Sprintf("%d", port), -1)
 }
+
+func (a *AppImpl) ServerNames(targetName string) []string {
+	// return from explicit targets, if exists
+	if _, ok := a.def.Targets[targetName]; ok {
+		var serverNames []string
+		return append(serverNames, targetName)
```
Comment (soulplant) [inline application.go]: Can just be written as `return []string{targetName}`

File: integration_test.go [hunk 4]
```diff
@@ -137,7 +138,12 @@ func startCamusWithConfig(t *testing.T, conf string) (*testClient, *os.Process)
 	time.Sleep(1 * time.Second)
 
 	isLocalTest := true
-	client, err := NewClientImpl(conf, "prod", isLocalTest)
+	isClient := true
+	app, err := ApplicationFromConfig(isClient, conf)
+	if err != nil {
+		t.Fatalf("NewClient:", err)
```
Comment (soulplant) [inline integration_test.go]: This should be `t.Fatalf("NewClient: %s", err)`

File: main.go [hunk 5]
```diff
@@ -53,11 +54,32 @@ func serverMain() {
 }
 
 func clientMain() {
-	client, err := NewClientImpl(*deployFile, *targetName, *isLocalTest)
+	isClient := true
+	app, err := ApplicationFromConfig(isClient, *deployFile)
 	if err != nil {
 		log.Fatal("NewClient:", err)
```
Comment (soulplant) [inline main.go]: This string also needs a %s in it.

File: main.go [hunk 6]
```diff
@@ -53,11 +54,32 @@ func serverMain() {
...
+	if serverNames == nil {
+		log.Fatal("No matching server target or deploy group:", *targetName)
+	}
+
+	// construct a client per server
+	var clients map[string]Client = make(map[string]Client)
+	for _, s := range serverNames {
+		c, err := NewClientImpl(app, path.Dir(*deployFile), s, *isLocalTest)
+		if err != nil {
+			log.Fatal("NewClient:", err)
```
Comment (soulplant) [inline main.go]: Needs %s

**Comment by jsuresh** [comment]
Abandoned this approach

**Pull request closed without merging**